flang/lib/Lower/IO.cpp

```diff
@@ -435,15 +435,23 @@ static void genIoLoop(Fortran::lower::AbstractConverter &converter,
   };
   if (!checkResult) {
     // No I/O call result checks - the loop is a fir.do_loop op.
-    auto loopOp =
-        builder.create<fir::DoLoopOp>(loc, lowerValue, upperValue, stepValue);
-    builder.setInsertionPointToStart(loopOp.getBody());
+    auto doLoopOp = builder.create<fir::DoLoopOp>(
+        loc, lowerValue, upperValue, stepValue, /*unordered*/ false,
+        ArrayRef<mlir::Value>{lowerValue}); // initial loop result value
+    builder.setInsertionPointToStart(doLoopOp.getBody());
     auto lcv = builder.createConvert(loc, converter.genType(loopSym),
-                                     loopOp.getInductionVar());
+                                     doLoopOp.getInductionVar());
     builder.create<fir::StoreOp>(loc, lcv, loopVar);
-    insertPt = builder.saveInsertionPoint();
     genItemList(ioImpliedDo);
-    builder.restoreInsertionPoint(parentInsertPt);
+    builder.setInsertionPointToEnd(doLoopOp.getBody());
+    mlir::Value result = builder.create<mlir::AddIOp>(
+        loc, doLoopOp.getInductionVar(), doLoopOp.step());
```

Comment:
Can't the  inductionVar + step Add happen after the loop if the loop returned the last induction variable value ?

Reply:
That might be a possibility for this DoLoopOp  case.  It wouldn't be completely the same for the IterWhileOp case.  There are some more options we could discuss.  It might be possible to have a DoLoopOp automatically define a result in terms of the surrogate loop control variable that is valid for ordered loops, but omitted for unordered loops.  That would avoid having the DoLoopOp generate any loads or stores, while still providing the correct value.  The IterWhileOp could do the same.  It would be even more useful there.  Currently an IterWhileOp isn't expected to have an ordered attribute, so that case could maybe have such a result all of the time, or it could also be conditional.

And if a DoLoopOp isn't really intended to directly reflect Fortran DO loop semantics, maybe we should name it something else.